### PR TITLE
General tidy-up to allow matching rules with targets

### PR DESCRIPTION
**Pull request by jayvdb**
**NOTE:** This PR includes the commits in https://github.com/EFForg/https-everywhere/pull/18938 - please lets get that PR merged first.

### List related issues if any

- #12297

I have a fairly functional tool for matching rules with targets in order to detect mismatches.

This PR includes some of the fixes for issues detected, and also regex syntax changes which make the matching easier.

Many of the fixes are in new smaller more focused PRs:

- https://github.com/EFForg/https-everywhere/pull/18949 - Broken ruleset deletions
- https://github.com/EFForg/https-everywhere/pull/18957 - Resolve error fixes

I appreciate that some of these syntax changes in this PR may create other problems that I am as yet unaware of -- I'm happy to drop parts of these changes and adjust my algorithm accordingly.

~~Scania.xml is one example where constraining the `\d+` is a "nice to have", but the change means that new hostnames brought online by Scania wouldnt automatically be included in the rule.  However it does mean that it is feasible for tools to grok the hostname sequence and generate tests dynamically, or detect if another hostname has been brought online, and check whether the regex needs updating.~~ (Scania.xml had bigger problems that this)

### CODE REVIEW

File: src/chrome/content/rules/VBSEO.xml [hunk 1]
```diff
@@ -13,6 +13,7 @@
 <ruleset name="vBSEO (partial)">
 
 	<target host="vbseo.com" />
+	<target host="cdn.vbseo.com" />
```
Comment (jayvdb) [inline src/chrome/content/rules/VBSEO.xml]: Created https://github.com/EFForg/https-everywhere/issues/18939

File: src/chrome/content/rules/Trusted_Reviews.xml [hunk 2]
```diff
@@ -23,6 +23,7 @@
 <ruleset name="Trusted Reviews (partial)">
 
 	<target host="trustedreviews.com" />
+	<target host="secure.trustedreviews.com" />
```
Comment (jayvdb) [inline src/chrome/content/rules/Trusted_Reviews.xml]: Moved to https://github.com/EFForg/https-everywhere/issues/18941

File: src/chrome/content/rules/polyvoreimg.com.xml [hunk 3]
```diff
@@ -20,18 +20,14 @@
...
 	<!--	Complications:
 				-->
-	<target host="ak1.polyvoreimg.com" />
-	<target host="ak2.polyvoreimg.com" />
 	<target host="cfc.polyvoreimg.com" />
 
 	<securecookie host="^\w" name=".+" />
 
 
-	<rule from="^http://(?:ak\d|akwww|cfc)\.polyvoreimg\.com/"
```
Comment (jayvdb) [inline src/chrome/content/rules/polyvoreimg.com.xml]: Moved to https://github.com/EFForg/https-everywhere/issues/18942

File: src/chrome/content/rules/raptrcdn.com.xml [hunk 4]
```diff
@@ -22,7 +22,7 @@
 	<target host="b1.raptrcdn.com" />
 
 
-	<rule from="^http://b\d\.raptrcdn\.com/"
+	<rule from="^http://b[01]\.raptrcdn\.com/"
```
Comment (jayvdb) [inline src/chrome/content/rules/raptrcdn.com.xml]: Moved to https://github.com/EFForg/https-everywhere/issues/18943

File: src/chrome/content/rules/Twenga.com.xml [hunk 5]
```diff
@@ -67,7 +67,7 @@
 	<securecookie host="^\." name="^(?:TW_VISITOR_ID|WSID|twr)" />
 
 
-	<rule from="^http://(blog|dynstatic|i\d+|r)\.twenga\.com/"
+	<rule from="^http://(blog|dynstatic|i(00|10|20|21|70|71)|r)\.twenga\.com/"
```
Comment (jayvdb) [inline src/chrome/content/rules/Twenga.com.xml]: Moved to https://github.com/EFForg/https-everywhere/issues/18944

File: src/chrome/content/rules/OceaniaCruises.com.xml [hunk 6]
```diff
@@ -17,6 +17,7 @@
 	<target host="l.discover.oceaniacruises.com" />
 	<target host="l.discover-intl.oceaniacruises.com" />
 	<target host="x.discover-intl.oceaniacruises.com" />
+	<target host="x.discover.oceaniacruises.com" />
```
Comment (jayvdb) [inline src/chrome/content/rules/OceaniaCruises.com.xml]: Moved to https://github.com/EFForg/https-everywhere/issues/18945

File: src/chrome/content/rules/Gartner.com.xml [hunk 7]
```diff
@@ -99,7 +99,7 @@
 	<securecookie host="^my\.gartner\.com$" name=".+" />
 
 
-	<rule from="^http://(?:na\d\.www\.)?gartner\.com/"
+	<rule from="^http://(?:na[1-3]\.www\.)?gartner\.com/"
```
Comment (jayvdb) [inline src/chrome/content/rules/Gartner.com.xml]: Moved to https://github.com/EFForg/https-everywhere/issues/18946

File: src/chrome/content/rules/UberTags.xml [hunk 8]
```diff
@@ -2,12 +2,11 @@
 -->
 <ruleset name="UberTags (partial)">
 
-	<target host="app.ubertags.com"/>
 	<target host="console.ubertags.com"/>
 
 	<securecookie host="^console\.ubertags\.com$" name=".+" />
 
-	<rule from="^http://(\w+)\.ubertags\.com/"
```
Comment (jayvdb) [inline src/chrome/content/rules/UberTags.xml]: Moved to https://github.com/EFForg/https-everywhere/pull/18949

Comment (jayvdb) [inline src/chrome/content/rules/UberTags.xml]: Moved to https://github.com/EFForg/https-everywhere/pull/18949

File: src/chrome/content/rules/SCEA.com.xml [hunk 9]
```diff
@@ -48,7 +48,7 @@
 	<!--rule from="^http://(?:www\.)?scea\.com/"
 		to="https://us.playstation.com/" /-->
 
-	<rule from="^http://webassets\w\.scea\.com/"
+	<rule from="^http://webassets[a-j]\.scea\.com/"
```
Comment (jayvdb) [inline src/chrome/content/rules/SCEA.com.xml]: moved to https://github.com/EFForg/https-everywhere/issues/18953

File: src/chrome/content/rules/Sinaimg.cn.xml [hunk 10]
```diff
@@ -175,7 +175,7 @@
 	<target host="s15.sinaimg.cn" />
 	<target host="s16.sinaimg.cn" />
 
-	<rule from="^http://s(\d+)\.sinaimg\.cn/" to="https://ss$1.sinaimg.cn/" />
+	<rule from="^http://s([1-9]|1[0-6])\.sinaimg\.cn/" to="https://ss$1.sinaimg.cn/" />
```
Comment (jayvdb) [inline src/chrome/content/rules/Sinaimg.cn.xml]: moved to https://github.com/EFForg/https-everywhere/issues/18954

File: src/chrome/content/rules/Bulletproof_Exec.com.xml [hunk 11]
```diff
@@ -41,7 +42,4 @@
 	<rule from="^http://(www\.)?bulletproofexec\.com/(?!/*(?:$|\?)|(?:category/podcasts|contact|have-dave-asprey-speak-at-your-next-event|upgraded-whey)(?:$|[?/]))"
 		to="https://$1bulletproofexec.com/" />
 
-	<rule from="^http://support\.bulletproofexec\.com/favicon\.ico"
-		to="https://bulletproofexec.zendesk.com/favicon.ico" />
```
Comment (jayvdb) [inline src/chrome/content/rules/Bulletproof_Exec.com.xml]: Moved to https://github.com/EFForg/https-everywhere/issues/18958

**Comment by jayvdb** [comment]
All the removals split off to https://github.com/EFForg/https-everywhere/pull/18957

**Comment by jayvdb** [comment]
CI has two consistent fetch errors that I can not reproduce here in Indonesia
```
ERROR src/chrome/content/rules/AliCDN.com.xml: Non-2xx HTTP code: http://amos.alicdn.com/ (200) => https://amos.alicdn.com/ (404)
ERROR src/chrome/content/rules/Qhimg.com.xml: Fetch error: http://weather.ssl.qhimg.com/ => https://weather.ssl.qhimg.com/: (52, 'Empty reply from server')
```
However I can reproduce one of the errors using wget, so it might be a user agent sniffer in play.
```
wget -S https://amos.alicdn.com/
--2020-02-20 11:09:13--  https://amos.alicdn.com/
Resolving amos.alicdn.com (amos.alicdn.com)... 203.119.169.207
Connecting to amos.alicdn.com (amos.alicdn.com)|203.119.169.207|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 404 Not Found
  Date: Thu, 20 Feb 2020 04:09:16 GMT
  Content-Type: text/html; charset=iso-8859-1
  Content-Length: 198
  Connection: keep-alive
  Server: Tengine/Aserver
2020-02-20 11:09:16 ERROR 404: Not Found.
```

File: src/chrome/content/rules/Affili.net.xml [hunk 12]
```diff
@@ -11,6 +11,8 @@
 
 	<target host="affili.net"/>
 	<target host="www.affili.net"/>
+	<target host="affili.de"/>
+	<target host="www.affili.de"/>
```
Comment (jayvdb) [inline src/chrome/content/rules/Affili.net.xml]: Moved to https://github.com/EFForg/https-everywhere/pull/18949

File: src/chrome/content/rules/SYW_CDN.net.xml [hunk 13]
```diff
@@ -17,7 +17,7 @@
 	<target host="*.sywcdn.net" />
 
 
-	<rule from="^http://(apps|\d+)\.sywcdn\.net/"
+	<rule from="^http://(apps|s(?:[1-9]|10))\.sywcdn\.net/"
```
Comment (jayvdb) [inline src/chrome/content/rules/SYW_CDN.net.xml]: Moved to https://github.com/EFForg/https-everywhere/issues/18963

**Comment by cschanaj** [comment]
@jayvdb Thank you for contributing to this repository. In general we don't recommend submitting huge PRs involving lots of unrelated rulesets. A special case is when the change is performed using a script/ tool which shall be peer reviewed and the outcome is reproducible. Otherwise, it is easier for reviewers to have a look on simpler PR and respond more promptly.

**Comment by jayvdb** [comment]
No fetch errors for AliCDN.com.xml on CI.

[CI](https://travis-ci.org/EFForg/https-everywhere/jobs/653126547) still has fetch errors for Qhimg.com.xml that I am not seeing locally.

> ERROR src/chrome/content/rules/Qhimg.com.xml: Fetch error: http://s9.qhimg.com/ => https://s.ssl.qhimg.com/: (52, 'Empty reply from server')

**Comment by jayvdb** [comment]
@cschanaj as noted in the PR description I am developing a tool.

I am most interested in feedback on the types of changes which would be unacceptable.  There is no point me perfecting a tool which has rules/makes changes that the maintainers considers unacceptable.

I also dont want to be managing 100+ PRs while I try to fix some of the worst *active* rulesets which have not been maintained for 5+ years and many fail basic sanity checks by CI which are skipped for these files unless someone touches them, and many have even been whitelisted.  I've avoided using the whitelist as a way to bypass the sanity checks, as development of my tool needs to understand the underlying problems and how to handle/report/fix them correctly.

These changes have a common theme of being extreme problems preventing the tool I am building from functioning sanely.  I am splitting them into smaller groups of changes when I understand how to group them better.

**Comment by jayvdb** [comment]
[CI 'fetch'](https://travis-ci.org/EFForg/https-everywhere/jobs/653289862) again errors on `Qhimg.com.xml`

> ERROR src/chrome/content/rules/Qhimg.com.xml: Fetch error: http://s2.qhimg.com/ => https://s2.ssl.qhimg.com/: (52, 'Empty reply from server')

Also frequently seeing

> ERROR src/chrome/content/rules/E-junkie.xml: Fetch error: http://static.e-junkie.com/manifest.json?v=5.0 => https://static.e-junkie.com/manifest.json?v=5.0: (28, 'Connection timed out after 20000 milliseconds')
> ERROR src/chrome/content/rules/E-junkie.xml: Fetch error: http://static.e-junkie.com/ => https://static.e-junkie.com/: (28, 'Connection timed out after 20001 milliseconds')

I encounter this today when running `check_rules.py` locally, but it wasnt problematic yesterday, and it works fine in a browser locally.

I guess these are good reasons to add a ruleset to the whitelist.

**Comment by cschanaj** [comment]
> I am most interested in feedback on the types of changes which would be unacceptable.

Usually, we prefer to have the source code for such script/ tools to be merged in the main repository first (then another PR for its outcome). So that reviewers can take a look and spot problems in the algorithm used. Personally, it is easier than manually reviewing a huge number of rulesets. In such scenario, we are relatively loose to fetch errors. If the reviewers agree, you can simply whitelist the rulesets failing fetch test instead of adding complication to the PR. 

cc @zoracon @J0WI @jeremyn @pipboy96 for comment.

**Comment by zoracon** [comment]
> I also dont want to be managing 100+ PRs while I try to fix some of the worst _active_ rulesets which have not been maintained for 5+ years and many fail basic sanity checks by CI which are skipped for these files unless someone touches them, and many have even been whitelisted. I've avoided using the whitelist as a way to bypass the sanity checks, as development of my tool needs to understand the underlying problems and how to handle/report/fix them correctly.
> 
@jayvdb 
On the point of having separate PRs I am noticing you are opening a lot of issues and pull requests consecutively. Thank you for your enthusiasm for the project, but I really suggest allowing the maintainers and I to review what you have already opened first and resolve those issues. I plan on running a [fetch test ](https://github.com/EFForg/https-everywhere/issues/18965) soon to help clean up a lot of these rulesets, but I am doing my best to create reviewable patches this time a round so that the maintainers can go over the results of the fetch test in a sane manner.

As frustrating opening separate PRs can can be, there is no directive to fix all problematic rulesets at once, since that is not reasonable of me to expect of volunteers or myself, and the maintainers work pretty fast when given a sane load. I ask that you trust that they will review your PRs in a sensible timeline.

For the errors you see that are intermittent, at times hosts can have outages and then be functional again at times. Which why we run the fetch test in 3 different time zones at 3 different times. To catch that sort of discrepancy.

File: src/chrome/content/rules/Delaware_Technical_Community_College.xml [hunk 14]
```diff
@@ -4,5 +4,7 @@
   <target host="www.dtcc.edu" />
 
   <rule from="^http://(?:www\.)?dtcc\.edu/" to="https://www.dtcc.edu/" />
-  <rule from="^http://(?:www\.)?login\.dtcc\.edu/" to="https://login.dtcc.edu/" />
+
+	<rule from="^http:"
```
Comment (thatch) [inline src/chrome/content/rules/Delaware_Technical_Community_College.xml]: This one doesn't look right.

File: src/chrome/content/rules/4sqi.net.xml [hunk 15]
```diff
@@ -17,7 +17,7 @@
 		<test url="http://ss3.4sqi.net/img/categories_v2/nightlife/beergarden_bg_88.png" />
 
 
-	<rule from="^http://(irs\d|is\d|ss[0-5l])\.4sqi\.net/"
+	<rule from="^http://(irs[-3]|is[0-5]|ss[0-5])\.4sqi\.net/"
```
Comment (thatch) [inline src/chrome/content/rules/4sqi.net.xml]: `[-3]` seems odd, in python at least that is either a dash or a 3, not a range.

File: src/chrome/content/rules/Fasthosts.xml [hunk 16]
```diff
@@ -39,7 +39,7 @@
 	<rule from="^http://(?:www\.)?fasthosts\.co\.uk/"
 		to="https://www.fasthosts.co.uk/" />
 
-	<rule from="^http://secure(\d|1[0-3-9]|2[0-4]|5[26])\.prositehosting\.co\.uk/"
```
Comment (thatch) [inline src/chrome/content/rules/Fasthosts.xml]: I'm not sure what the original `[0-3-9]` was intended to match; the replacement pattern is not what python currently matches (0-3, or dash, or 9).

Comment (pipboy96) [reply src/chrome/content/rules/Fasthosts.xml]: @thatch This is correct. This will match any of `['0', '1', '2', '3', '9', '-']`.

Comment (thatch) [reply src/chrome/content/rules/Fasthosts.xml]: @pipboy96 the replacement on the line below doesn't escape the `-` in `-9` though, so it changes the meaning.

File: src/chrome/content/rules/4sqi.net.xml (hunk 15 above)
Comment (thatch) [reply src/chrome/content/rules/4sqi.net.xml]: Related to the `0-3-9` comment below, dash handling needs some work.

**Comment by jayvdb** [comment]
Thanks for all the merges; I'll go through and rebase/discard whatever is left on the w/e.

**Pull request closed without merging**